### DISCUSSION TITLE
Menu's replaced with Popups

### DISCUSSION THREAD

**Discussion by drmikecrowe** [skeletonlabs/skeleton#1043] [Q&A]
My timing was impeccable -- heard @endigo9I740 on Podrocket, started using Skeleton, built my Menu, then they went away :smiley: 

I agree the foundation of Popup does everything we need for a menu, but unless I'm missing something, I feel like you can't remove Menu and call it an RC.  In order to replicate the prior functionality, I had to pull in this from the docs site:

```ts
	import { popup } from '$lib/utilities/Popup/popup'
	import type { PopupSettings } from '$lib/utilities/Popup/types'
```

I'm not sure what in this is required, but here's what I used before -- I really feel the base expectation is this should be supported out of the box:

```svelte
<div>
	{#if $fullUserStore && $fullUserStore.ownedDevices.length}
		<span class="relative">
			<button class="btn bg-primary-500" use:menu={{ menu: 'example' }}>{selectedDevice}</button>
			<nav class="list-nav card p-4 w-64 shadow-xl" data-menu="example">
				<ListBox data-menu="example" class="text-left">
					{#each $fullUserStore.ownedDevices as item}
						<ListBoxItem bind:group={selectedDevice} name="items" value={item}>{item}</ListBoxItem>
					{/each}
				</ListBox>
			</nav>
		</span>
	{/if}
</div>
```

**Comment by drmikecrowe** [comment]
Also, from a documentation perspective, the Popup documentation needs to state that the `+layout.svelte` changes must be at the top of the script (prior to importing the component).  Otherwise, the popup fails.

**Comment by endigo9740** [comment]
> I agree the foundation of Popup does everything we need for a menu, but unless I'm missing something, I feel like you can't remove Menu and call it an RC. In order to replicate the prior functionality, I had to pull in this from the docs site:

It literally does everything the menu does, and does it better in nearly every regard. We consider it a complete replacement. Keeping the menu would have introduced a redundancy with issues that we will not be supporting any longer. We're calling it a release candidate because it is a release candidate. You don't have to agree with that.

> I'm not sure what in this is required, but here's what I used before -- I really feel the base expectation is this should be supported out of the box:

Everything you need to use the popups is shared on the Popup documentation page. Read that top-to-bottom and you'll have your answers for how to use it. That's why we spend the time writing up these documents in the first place:
https://www.skeleton.dev/utilities/popups

> Also, from a documentation perspective, the Popup documentation needs to state that the +layout.svelte changes must be at the top of the script (prior to importing the component). Otherwise, the popup fails.

We do mention it's required to go in the root layout:

![image](https://user-images.githubusercontent.com/1509726/221372068-a2274314-8ad7-4143-b367-d60030d37b26.png)

There's no component involved with the Popup. It's a Svelte action.

You import the Floating UI dependency, you pass a reference to the store (which keeps it option), then the action utilizes this instance. If you fail to follow the provided instruction or skip steps things aren't going to work. We literally write the instructions for this in the order needed top to bottom. If folks opt not to follow that we really have no control over this.

Just as an aside here, we do implement a Code of Conduct in our community. We ask that feedback be constructive and empathetic. I'm just going to let you know the language used here is riding the line. I'm going to go ahead and lock this thread now that your issues have been heard and addressed. But please consider finding another way to approach similar discussions in the future:
https://github.com/skeletonlabs/community/blob/main/CODE_OF_CONDUCT.md

**Comment by endigo9740** [comment]
@drmikecrowe I've decided to reverse the lock so you can respond. We do welcome open discourse here, so I apologize for removing the option for that.

We're fast approaching our v1 release which means stress levels are at max right now. I apologize if my message was not constructive. If we can help with your issues above let us know.

**Reply by drmikecrowe** [reply]
Thanks @endigo9740 for letting me respond. I feel like I'm a fairly typical user, so I'm a bit confused as to my observation might have been considered non-constructive. I'd appreciate any feedback as to how I could have better phrased it in the future, cuz my intent was certainly to be positive.

When I reviewed the documentation, what I felt like I saw was something that required quite a bit of utility functions in order to implement the drop-down menu. For me, the prior menu component seem to implement exactly what I wanted, whereas I had to dig into the documentation source to pull out the use code to implement what I was trying to achieve. Granted, this could totally be ignorance on my part, but it felt like there was a gap, so I raised this as feedback from a new user's perspective.

**Reply by endigo9740** [reply]
For context, Skeleton is an open source project which means all work happens publicly. Between tickets, PRs, and discussion on Discord. No changes are made in secret. I always always ask for feedback for changes before the work begins, while it's in progress, or during the testing phase leading to release.

However, very few folks opt to participate in this process. Your comment about this "shouldn't be a release candidate" honestly rubbed me the wrong way. You haven't participated in the conversation for this, which means you may be (unintentionally) ignorant to the reasons why this has occurred. So I took issue with that.

Just FYI we've been following a roadmap for months. Our release candidate and v1 are based on certain milestones being reached. This was not done willy nilly. This is on consensus of myself and the core team. 

I would just encourage you get caught up with things before making snap judgment like this. And remember this is being provided completely for free to you and others. Contributed by volunteers that do this for no compensation. Some of us working 7 days a week and upwards of 14 hours a day. It's no small effort to make this happen, so all we ask for in return is to keep feedback kind and constructive. Hope that makes sense.

**Reply by endigo9740** [reply]
Here's the roadmap btw: https://github.com/skeletonlabs/skeleton/discussions/503

**Reply by drmikecrowe** [reply]
@endigo9740 thank you, that was exactly what I needed. I did join the late, and did not participate, and that is very good constructive feedback. 

Let me experiment some more and understand better how to use the pop-up to keep the menu that I was looking for. 

Since I got confused over the documentation, is that something I could offer a PR once I work through my problem? Maybe I can submit an example from a new user's perspective that illustrates the basic functionality I was looking for but didn't immediately find. 

Regarding the placement of the imports in layout, my issue was I placed the pop-up imports at the bottom of the layout import list, and had imported the menu component above these new lines, which caused my error. I think the clarification for me would simply be in the layouts section, a simple 'include the following imports early in your layout.svelte, prior to any imports using the pop-up component functionality'. 

Let me know if you would welcome documentation we are, and I can submit some to assist. That goes well, I'll be happy to ship in some more and become a better contributor in the future.

**Reply by riziles** [reply]
> Let me know if you would welcome documentation we are, and I can submit some to assist. That goes well, I'll be happy to ship in some more and become a better contributor in the future.

@drmikecrowe , are you on the Discord server? There's a lot of great dialog on there. I highly recommend joining if you aren't already on it.

**Reply by endigo9740** [reply]
@drmikecrowe No worries at all, honestly feedback is exactly what we're needing so what you provided is valuable. You can note here we're actively in progress on major improvements to the documentation right now. I'm super excited to see folks reaction to what we have in the works:
- https://github.com/skeletonlabs/skeleton/issues/783

I've added a quick note to address some of the gaps you've highlighted above. One-off PRs are great, but I think we're better served to wait until me and several other volunteers do our focused dive through this over the course of the next week.

If you would like to be a part of that group feel free to join Discord as Riziles mentioned. See the `#doc-overhaul` channel on there for details. I'll be finalizing the last few prereqs tomorrow that will allow that process to begin.

**Comment by riziles** [comment]
@drmikecrowe , what does RC stand for in this context?

**Reply by endigo9740** [reply]
Release Candidate

**Reply by riziles** [reply]
Oh.... I see now
